cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -397,181 +399,191 @@ func (m *McmManager) Cleanup() {
 	}
+	// update priority of all machines in the machineDeployment to 3 except the ones that have nodes with the ToBeDeleted taint
+	if err = m.resetPriorityForNotToBeDeletedMachines(commonMachineDeployment.Name); err != nil {
+		return err
+	}
+	// update priorities of machines to be deleted except the ones already in termination to 1
+	scaleDownAmount, err := m.prioritizeMachinesForDeletion(targetMachineRefs, commonMachineDeployment.Name)
+	if err != nil {
+		return err
+	}
+	// Trying to update the machineDeployment till the deadline
+	err = m.retry(func(ctx context.Context) (bool, error) {
+		return m.scaleDownMachineDeployment(ctx, commonMachineDeployment.Name, scaleDownAmount)
+	}, "MachineDeployment", "update", commonMachineDeployment.Name)
+	if err != nil {
+		klog.Errorf("unable to scale in machine deployment %s, err:%v", commonMachineDeployment.Name, err)
+		return fmt.Errorf("unable to scale in machine deployment %s, err: %v", commonMachineDeployment.Name, err)
+	}
+	return nil
+}
 
-	for _, machine := range machines {
-
-		// Trying to update the priority of machine till retryDeadline
-		retryDeadline := time.Now().Add(maxRetryTimeout)
-		for {
-			machine, err := m.machineLister.Machines(m.namespace).Get(machine.Name)
-			if err != nil && time.Now().Before(retryDeadline) {
-				klog.Warningf("Unable to fetch Machine object %s, Error: %s , will retry in %s", machine.Name, err, conflictRetryInterval)
-				time.Sleep(conflictRetryInterval)
-				continue
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to fetch Machine object %s, Error: %s", machine.Name, err)
-				return err
-			}
-
-			mclone := machine.DeepCopy()
-
-			if isMachineTerminating(mclone) {
-				terminatingMachines = append(terminatingMachines, mclone)
-			} else {
-				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mclone.Name, mclone.Labels["node"]})
-			}
-
-			if mclone.Annotations != nil {
-				if mclone.Annotations[machinePriorityAnnotation] == "1" {
-					klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
-					break
+// resetPriorityForNotToBeDeletedMachines resets the priority of machines with nodes without ToBeDeleted taint to 3
+func (m *McmManager) resetPriorityForNotToBeDeletedMachines(mdName string) error {
+	allMachinesForMachineDeployment, err := m.getMachinesForMachineDeployment(mdName)
+	if err != nil {
+		return fmt.Errorf("unable to list all machines for node group %s, cannot proceed with scale-in of machines, Error: %v", mdName, err)
+	}
+	for _, machine := range allMachinesForMachineDeployment {
+		val, ok := machine.Annotations[machinePriorityAnnotation]
+		if ok && val != defaultPriorityValue {
+			if err := m.retry(func(ctx context.Context) (bool, error) {
+				nodeName := machine.Labels[v1alpha1.NodeLabelKey]
```

Comment:
I think this can be solved by changing the `defaultMaxRetryTimeout`. We can keep it as it is for now and change it if we encounter any problems. Also, as soon as one machine encounters an error, we will return the error instead of trying for other machines. So in case of API server not responding, the return time will be `retryTimeout`.

Reply:
I am not sure that is the right behavior. Shouldn't you try and update the annotations for as many machines as possible and then return a collective error? I am still not sure if retry should be at the machine level.

Reply:
I don't see an advantage in using collective error in this case. The Refresh method will return an error even if one update call fails. So, in any case, another reconciliation will be needed. If the API server is down, all the update calls will fail with the same error and returning a collective error in the logs might make them bloated.

Reply:
In the refresh method shouldn't be a best effort to reset priority annotation value to 3 for as many machines as possible? This way every run of `refresh` method will then only deal with the remaining ones that could not be updated before. in the current case lets say the first update always fails for some reason, then no other machine will ever get the annotation value changed. 

Collective error is only meaningful if you provide best effort update of annotation on machines. This means that you go on even if there are some machines which could not be updated. A simple use of `errors.Join` would suffice here.

Reply:
Ok for now we have decided to return an error from `resetPriorityForNotToBeDeletedMachines` knowing that any error returned when updating annotation on machines will cause the `RunOnce` flow to be aborted and retried after 10s. Any need for a scale-up which is handled post call to `Refresh` will also get delayed. Resetting annotations is only required during a scale-down flow and the side effect of using `Refresh` method is that failure to do that would impact the overall `RunOnce` reconciliation flow.

We should watch for issues due to this and make changes if required later.